File: ios/Classes/Bridges/SubscriptionStatusBridge.swift
import Flutter
import SuperwallKit

// Abstract base class for subscription status enum
public class SubscriptionStatusBridge: BridgeInstance {
  var status: SubscriptionStatus {
    assertionFailure("Subclasses must implement")
    return .unknown
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getDescription":
      let description = status.description
      result(description)
    case "getEntitlements":
      let entitlements = Superwall.shared.entitlements.active.map { $0.toJson() }
      result(entitlements)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

public class SubscriptionStatusActiveBridge: SubscriptionStatusBridge {
  class override var bridgeClass: BridgeClass { "SubscriptionStatusActiveBridge" }
  override var status: SubscriptionStatus { .active(entitlements) }

  let entitlements: Set<Entitlement>

  required init(
    bridgeId: BridgeId,
    initializationArgs: [String: Any]? = nil
  ) {
    if let entitlementsJson = initializationArgs?["entitlements"] as? [[String: Any]] {
      self.entitlements = Set(entitlementsJson.compactMap { Entitlement.fromJson($0) })
    } else {
      self.entitlements = []
    }

    super.init(bridgeId: bridgeId, initializationArgs: initializationArgs)
  }
}

public class SubscriptionStatusInactiveBridge: SubscriptionStatusBridge {
  class override var bridgeClass: BridgeClass { "SubscriptionStatusInactiveBridge" }
  override var status: SubscriptionStatus { .inactive }
}

public class SubscriptionStatusUnknownBridge: SubscriptionStatusBridge {
  class override var bridgeClass: BridgeClass { "SubscriptionStatusUnknownBridge" }
  override var status: SubscriptionStatus { .unknown }
}

extension SubscriptionStatus {
  func createBridgeId() -> BridgeId {
    switch self {
    case .active:
      let bridgeInstance = BridgingCreator.shared.createBridgeInstance(
        bridgeClass: SubscriptionStatusActiveBridge.bridgeClass)
      return bridgeInstance.bridgeId
    case .inactive:
      let bridgeInstance = BridgingCreator.shared.createBridgeInstance(
        bridgeClass: SubscriptionStatusInactiveBridge.bridgeClass)
      return bridgeInstance.bridgeId
    case .unknown:
      let bridgeInstance = BridgingCreator.shared.createBridgeInstance(
        bridgeClass: SubscriptionStatusUnknownBridge.bridgeClass)
      return bridgeInstance.bridgeId
    }
  }
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/bridges/SubscriptionStatusBridge.kt
package com.superwall.superwallkit_flutter.bridges

import android.content.Context
import android.util.Log
import com.superwall.sdk.models.entitlements.SubscriptionStatus
import com.superwall.sdk.Superwall
import com.superwall.superwallkit_flutter.BridgingCreator
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import com.superwall.superwallkit_flutter.json.toEntitlement
import com.superwall.superwallkit_flutter.json.toJson

open class SubscriptionStatusBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : BridgeInstance(context, bridgeId, initializationArgs) {
    open val status: SubscriptionStatus
        get() = throw AssertionError("Subclasses must implement")

    override val cachable = false
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getDescription" -> {
                val description = status.toString()
                result.success(description)
            }
            "getEntitlements" -> {
                val entitlements = Superwall.instance.entitlements.active.map { it.toJson() }.toList()

                result.success(entitlements)
            }
            else -> result.notImplemented()
        }
    }
}

// Define the subclasses for each subscription status
class SubscriptionStatusActiveBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>?
) : SubscriptionStatusBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "SubscriptionStatusActiveBridge" }


    override val status: SubscriptionStatus
        get() {
            Log.e("REceived entitlements", "$initializationArgs")
            val entitlements = (initializationArgs?.get("entitlements") as? List<Map<String,Any>>)?.map {
                Log.e("Deserializing", "To ent: $it")
                it.toEntitlement() }?.toSet() ?: emptySet()
            return SubscriptionStatus.Active(entitlements = entitlements)
        }
}

class SubscriptionStatusInactiveBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : SubscriptionStatusBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "SubscriptionStatusInactiveBridge" }

    override val status: SubscriptionStatus
        get() = SubscriptionStatus.Inactive
}

class SubscriptionStatusUnknownBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : SubscriptionStatusBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "SubscriptionStatusUnknownBridge" }

    override val status: SubscriptionStatus
        get() = SubscriptionStatus.Unknown
}

suspend fun SubscriptionStatus.createBridgeId(): BridgeId {
    val bridgeClass = when (this) {
        is SubscriptionStatus.Active -> SubscriptionStatusActiveBridge.bridgeClass()
        is SubscriptionStatus.Inactive -> SubscriptionStatusInactiveBridge.bridgeClass()
        is SubscriptionStatus.Unknown -> SubscriptionStatusUnknownBridge.bridgeClass()
    }

    val bridgeInstance = BridgingCreator.shared.createBridgeInstanceFromBridgeClass(bridgeClass)
    return bridgeInstance.bridgeId
}